src/Microsoft.AspNet.Mvc.Xml/DelegatingEnumerator.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNet.Mvc.Xml
+{
+    /// <summary>
+    /// Delegates enumeration of elements to the original enumerator and wraps the items
+    /// with the supplied <see cref="IWrapperProvider"/>.
+    /// </summary>
+    /// <typeparam name="TWrapped">The type to which the individual elements need to be wrapped to.</typeparam>
+    /// <typeparam name="TDeclared">The original type of the element being wrapped.</typeparam>
+    public class DelegatingEnumerator<TWrapped, TDeclared> : IEnumerator<TWrapped>
+    {
+        private readonly IEnumerator<TDeclared> _inner;
+        private readonly IWrapperProvider _wrapperProvider;
+
+        /// <summary>
+        /// Initializes a <see cref="DelegatingEnumerable{TWrapped, TDeclared}"/> which enumerates 
+        /// over the elements of the original enumerator and wraps them using the supplied
+        /// <see cref="IWrapperProvider"/>.
+        /// </summary>
+        /// <param name="inner">The original enumerator.</param>
+        /// <param name="wrapperProvider">The wrapper provider to wrap individual elements.</param>
+        public DelegatingEnumerator([NotNull] IEnumerator<TDeclared> inner, IWrapperProvider wrapperProvider)
```

Comment:
Discussed with @kichalla - at this point in the code we already know what the inner-type's wrapper provider is, so we don't need the factory anymore.